File: api/internal/reviewai/observation.go
package reviewai

import (
	"regexp"
	"strings"
)

type Observation struct {
	JobID         string    `json:"job_id"`
	StepKey       string    `json:"step_key"`
	CurrentURL    string    `json:"current_url"`
	PageTitle     string    `json:"page_title"`
	VisibleText   string    `json:"visible_text"`
	DOMHints      []DOMHint `json:"dom_hints"`
	ScreenshotRef string    `json:"screenshot_ref,omitempty"`
}

type DOMHint struct {
	Role         string `json:"role"`
	Text         string `json:"text"`
	SelectorHint string `json:"selector_hint"`
}

var secretPatterns = []*regexp.Regexp{
	regexp.MustCompile(`(?i)(password|passcode|verification code|2fa|token|secret|api key)\s*[:=]?\s*\S+`),
	regexp.MustCompile(`sk-ant-[A-Za-z0-9_-]+`),
	regexp.MustCompile(`Bearer\s+[A-Za-z0-9._-]+`),
}

func RedactObservation(obs Observation) Observation {
	obs.VisibleText = RedactVisibleText(obs.VisibleText)
	obs.DOMHints = RedactDOMHints(obs.DOMHints)
	return obs
}

func RedactVisibleText(value string) string {
	out := value
	for _, pattern := range secretPatterns {
		out = pattern.ReplaceAllString(out, "[redacted]")
	}
	return out
}

func RedactDOMHints(hints []DOMHint) []DOMHint {
	out := make([]DOMHint, 0, len(hints))
	for _, hint := range hints {
		joined := strings.ToLower(hint.Role + " " + hint.Text + " " + hint.SelectorHint)
		if strings.Contains(joined, "password") ||
			strings.Contains(joined, "verification code") ||
			strings.Contains(joined, "2fa") {
			continue
		}
		hint.Text = RedactVisibleText(hint.Text)
		out = append(out, hint)
	}
	return out
}

File: review-agent/src/observation.js
const SECRET_PATTERNS = [
  /(password|passcode|verification code|2fa|token|secret|api key)\s*[:=]?\s*\S+/gi,
  /sk-ant-[A-Za-z0-9_-]+/g,
  /Bearer\s+[A-Za-z0-9._-]+/g,
];

export async function collectPageObservation(page, { jobId = "", stepKey = "" } = {}) {
  const currentUrl = typeof page.url === "function" ? page.url() : "";
  const title = await safePageTitle(page);
  const texts = await page
    .locator("body")
    .allInnerTexts()
    .catch(() => []);
  const domHints = await page
    .evaluate(() => {
      return Array.from(document.querySelectorAll("[data-review-step],button,a,input,textarea,select"))
        .filter((el) => {
          const rect = el.getBoundingClientRect();
          return rect.width > 0 && rect.height > 0;
        })
        .slice(0, 80)
        .map((el) => {
          const reviewStep = el.getAttribute("data-review-step") || "";
          return {
            role: el.getAttribute("role") || el.tagName.toLowerCase(),
            text: (el.innerText || el.value || el.getAttribute("aria-label") || el.getAttribute("placeholder") || "")
              .trim()
              .slice(0, 160),
            selector_hint: reviewStep ? `[data-review-step='${reviewStep}']` : "",
          };
        });
    })
    .catch(() => []);

  return redactObservation({
    job_id: jobId,
    step_key: stepKey,
    current_url: currentUrl,
    page_title: title,
    visible_text: texts.join("\n").slice(0, 12000),
    dom_hints: domHints,
  });
}

export function redactObservation(observation = {}) {
  return {
    ...observation,
    visible_text: redactText(observation.visible_text || ""),
    dom_hints: (observation.dom_hints || [])
      .filter((hint) => {
        const joined = `${hint.role || ""} ${hint.text || ""} ${hint.selector_hint || ""}`.toLowerCase();
        return !joined.includes("password") && !joined.includes("verification code") && !joined.includes("2fa");
      })
      .map((hint) => ({ ...hint, text: redactText(hint.text || "") })),
  };
}

async function safePageTitle(page) {
  if (typeof page.title !== "function") {
    return "";
  }
  try {
    return await page.title();
  } catch {
    return "";
  }
}

function redactText(value) {
  let out = String(value || "");
  for (const pattern of SECRET_PATTERNS) {
    out = out.replace(pattern, "[redacted]");
  }
  return out;
}
